src/sage/combinat/partitions.pyx

```diff
@@ -45,52 +124,129 @@ def ZS1_iterator(int n):
+
+    .. SEEALSO::
+
+        - :meth:`sage.combinat.partitions.ZS2_iterator`
+        - :meth:`sage.combinat.partitions.AccelDesc_iterator`
+
+    EXAMPLES::
+
+        sage: from sage.combinat.partitions import ZS1_iterator
+        sage: for p in ZS1_iterator(4):
+        ....:     print(p)
+        [4]
+        [3, 1]
+        [2, 2]
+        [2, 1, 1]
+        [1, 1, 1, 1]
+        sage: next(ZS1_iterator(4))
+        [4]
+        sage: type(_)
+        <class 'list'>
+    """
     # Easy cases.
     if n < 0:
         return
-    if n == 0:
+    if not n:
         yield []
         return
-    x = [1]*n
-    x[0] = n
-
+    cdef list P = [1] * n
+    P[0] = n
     cdef int m = 0
     cdef int h = 0
-    cdef int r, t
     yield [n]
-    while x[0] != 1:
-        # Loop invariants at this point:
-        # (A) x[:m+1] is a partition of n.
-        # (B) x[h+1:] is an array of n-(h+1) ones.
-        # (C) x[i] > 1 for each i <= h.
-        # (D) 0 <= h <= m.
-        if x[h] == 2:
-            m += 1
-            x[h] = 1
-            h -= 1
-        else:
-            t = m - h + 1
-            r = x[h] - 1
-            x[h] = r
-            while t >= r:
-                h += 1
-                x[h] = r
-                t -= r
-            if t == 0:
-                m = h
-            else:
-                m = h + 1
-                if t > 1:
-                    h += 1
-                    x[h] = t
-        # yield [x[i] for i in range(m+1)]
-        yield x[:m+1]
-    # free(x)
+    while P[0] != 1:
+        ZS1_step(P, n, &m, &h)
+        yield P[:m+1]
 
 
-def ZS1_iterator_nk(int n, int k):
+def ZS1_next(list P):
+    r"""
+    Return the partition after `P` in the ordering of the ZS1 algorithm.
+
+    INPUT:
+
+    - ``P`` -- a list encoding a partition of an integer `n` in descending order
+      (i.e., `P_i \geq P_{i+1}`)
+
+    EXAMPLES::
+
+        sage: from sage.combinat.partitions import ZS1_iterator, ZS1_next
+        sage: P = [4]
+        sage: while P:
+        ....:     print(P)
+        ....:     P = ZS1_next(P)
+        [4]
+        [3, 1]
+        [2, 2]
+        [2, 1, 1]
+        [1, 1, 1, 1]
+        sage: A = [list(p) for p in Partitions(7)]
+        sage: all(ZS1_next(p) == q for p, q in zip(A, A[1:]))
+        True
     """
-    An iterator for the partitions of ``n`` of length at most ``k`` (in the
+    if P[0] == 1:
+        return
+    if P[-1] != 1:
+        return P[:-1] + [P[-1] - 1, 1]
+
+    cdef int n = sum(P)
+    cdef int m = len(P) - 1  # index of the last element in P
+    # Search for the index h of the last value > 1 in P
+    cdef int h = m
+    while P[h] == 1:
+        h -= 1
+    # Let Q be such that Q[:m+1] == P and Q[h+1:] is an array of n-(h+1) ones
+    cdef list Q = P + [1] * (n - m - 1)
+    ZS1_step(Q, n, &m, &h)
+    return Q[:m+1]
+
+
+def ZS1_iterator_nk(int n, int k):
+    r"""
+    An iterator for the partitions of `n` of length at most `k` (in the
```

Comment:
```suggestion
    An iterator for the partitions of ``n`` of length at most ``k`` (in the
```
and similar changes since the are inputs (and hence "code").

Reply:
I think I changed all occurrences.